### ISSUE TITLE
AssertionError: results in not an array of object

### ISSUE THREAD

**Issue by VincentDuf** [pdollar/coco#401]
Hi,
I have an issue that I don't get when I want to use the **pycocoEvalDemo** jupyter notebook.
When I want to load the results file I get this error:
```
Loading and preparing results...

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-10-d6d4841cb710> in <module>
      1 #initialize COCO detections api
      2 resFile='/home/greenshield/CNN/valid_detect.json'
----> 3 cocoDt=cocoGt.loadRes(resFile)

~/coco/PythonAPI/pycocotools/coco.py in loadRes(self, resFile)
    321             anns = resFile
    322         assert type(anns) == list, 'results in not an array of objects'
--> 323         annsImgIds = [ann['image_id'] for ann in anns]
    324         assert set(annsImgIds) == (set(annsImgIds) & set(self.getImgIds())), \
    325                'Results do not correspond to current coco set'

AssertionError: results in not an array of objects
```

But my result file has been generated correctly and it respects the same format as COCO.
It looks like its coming from the line 322 where it does:
`assert type(anns)==list`
The problem is the ResFile is first of all a dictionnary containing list of dictionnaries. So I tried to put:
`assert type(anns['annotations']` 
or 
`anns=resFile['annotations']`
But in both case it doesn't take the change in account and i can't do a 'make clean', it tells me there is no procedure for that. 

What should I do to clear the make and re-do it to take the change into account ? 

Thanks

**Comment by bot66** [comment]
me too ！ why ? isn't load-in json file suppose to be a dict ?

**Comment by ljmd21** [comment]
Same here! Has anyone solved this? Thank you!!

**Comment by pradeep-wyze** [comment]
Anyone solved it?

**Comment by Hiteshsaai** [comment]
Did any one solve this issue ??

**Comment by SkalskiP** [comment]
Any updates on that?

**Comment by zxq309** [comment]
Same Problem Sadly! Can Anyone Solve this?

**Comment by SkalskiP** [comment]
@XiangqingZhang `pycocotools` is not developed for the last 3 years. 😒  So I know it is sad but, I don't think any fix will happen. I managed to control the problem:

Try to install `pycocotools` first and then install `numpy`. In that case order matters.

**Comment by BatmanofZuhandArrgh** [comment]
Apparently, coco prediction json has a different format than ground truth json. It should be like this:
[
    {
        "image_id": 1,
        "category_id": 0,
        "bbox": [
            466.484375,
            76.4969711303711,
            224.9879150390625,
            94.5906753540039
        ],
        "score": 0.9025385975837708,
        "segmentation": []
    },
    {
        "image_id": 1,
        "category_id": 0,
        "bbox": [
            134.98641967773438,
            230.1739044189453,
            173.26910400390625,
            321.313720703125
        ],
        "score": 0.8995603919029236,
        "segmentation": []
    },
    {
        "image_id": 1,
        "category_id": 1,
        "bbox": [
            684.7588500976562,
            110.65011596679688,
            31.15533447265625,
            42.84550476074219
        ],
        "score": 0.88946133852005,
        "segmentation": []
    },
...
]

**Comment by ZahraAnam** [comment]
Well The problem is that its type is dict but is asserted to be list
which is technically wrong - coz .json loads as .dict not as list

**Comment by BatmanofZuhandArrgh** [comment]
@ZahraAnam  there are 2 types to coco format:
- Data format: https://cocodataset.org/#format-data
- Result format: https://cocodataset.org/#format-results
The result format after loaded is list

**Comment by VtojeC** [comment]
I have this error too and my files are in the format @BatmanofZuhandArrgh suggested so... @ZahraAnam has a point.
I don't know what is a solution to this problem.

Have you solved that issue? I'd be grateful for a help.

EDIT:
- I think I solved it. My file formats are okay, but my GT file was having 1 less photo than Result File so program was crying about it as it was looking like results didn't correspond to current coco set.

IN SHORT:
Result file had 1 more image that Ground Truth file didn't so program was not prepared to deal with it.

**Comment by ZahraAnam** [comment]
I generally change the output file to adapt for comparison in Coco Evaluation in MatlabAPI

**Comment by gguzzy** [comment]
Hi dear all,
I had the same issue, and I solved using this specific script I created, if you want to look it up:
[https://github.com/gguzzy/coco-annotation-converter/blob/main/convert_coco_data_to_result](url)
This script converts from coco data format to result format, allowing us to properly run evaluation scripts from coco.

If you want to insert it as a merge request to cocoapi, please contact me!

I hope it helps,
thanks!

**Comment by lesept777** [comment]
> Hi dear all, I had the same issue, and I solved using this specific script I created, if you want to look it up: [https://github.com/gguzzy/coco-annotation-converter/blob/main/convert_coco_data_to_result](url) This script converts from coco data format to result format, allowing us to properly run evaluation scripts from coco.
> 
> If you want to insert it as a merge request to cocoapi, please contact me!
> 
> I hope it helps, thanks!

The link is void

**Comment by gguzzy** [comment]
Hi sorry for late reply, please check the link in here [convert_coco_data_to_result](https://github.com/gguzzy/coco-annotation-converter ). Thanks, hope you will fix that!

**Comment by adiv5** [comment]
This worked for me
```
from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval
import json

# Load ground truth (val.json)
coco_gt = COCO("path_to_gt_json")

with open("path_to_prediction_json", 'r') as f:
    predictions = json.load(f)
    
# Load predictions 
coco_dt = coco_gt.loadRes(predictions)  # Pass the prediction list instead of filepath to prediction JSON

# Initialize COCO evaluator
coco_eval = COCOeval(coco_gt, coco_dt, "bbox")

# Run evaluation
coco_eval.evaluate()
coco_eval.accumulate()
coco_eval.summarize()
```
